tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php

```diff
@@ -475,7 +477,7 @@ public function testPrepareQueryBindValueDateTimeType() : void
     {
         $sql  = 'SELECT count(*) AS c FROM fetch_table WHERE test_datetime = ?';
         $stmt = $this->connection->prepare($sql);
-        $stmt->bindValue(1, new DateTime('2010-01-01 10:10:10'), Types::DATETIME_MUTABLE);
+        $stmt->bindValue(1, '2010-01-01 10:10:10', ParameterType::STRING);
```

Comment:
What type should it be? I got an error about not being able to cast DateTime to a string.

Reply:
`bindValue()` of a driver statement supports only `int` type but the DBAL statement which implements driver statement (which is a big question) supports also supports at least `string` and `Type` as `$type`. PHPStan should be given proper info on the statement type (dbal statement).

As for the value, all implementations accept `mixed $value`, so it must be some mistake.

Reply:
This is the error when I try to bind a DateTime:

```
1) Doctrine\Tests\DBAL\Functional\DataAccessTest::testPrepareQueryBindValueDateTimeType
Object of class DateTime could not be converted to string

/data/repositories/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:75
/data/repositories/dbal/lib/Doctrine/DBAL/Statement.php:113
/data/repositories/dbal/tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php:480
```

What is supposed to be responsible for converting the DateTime?

Reply:
It should be here and it seems to have worked before these changes:

https://github.com/doctrine/dbal/blob/d07906fddf156df4a3de8df28764e1485e4c885c/lib/Doctrine/DBAL/Statement.php#L102-L111

Reply:
The API does not allow this to happen. This used to be possible:

```php
$stmt->bindValue(1, new DateTime('2010-01-01 10:10:10'), Types::DATETIME_MUTABLE);
```

Now that we enabled phpstan in the tests, we get this error:

```
 ------ -----------------------------------------------------------------------------------------------------
  Line   tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php
 ------ -----------------------------------------------------------------------------------------------------
  480    Parameter #3 $type of method Doctrine\DBAL\Driver\Statement::bindValue() expects int, string given.
 ------ -----------------------------------------------------------------------------------------------------
```

Is the API wrong, or do we need to always pass an `int` constant from `ParameterType` now?

Reply:
It's the same problem as with the previous assertion. PHPStan thinks you have a driver statement which only accepts `int` as a parameter while you're dealing with a DBAL statement which accepts other types too.

Reply:
Ok, I understand now. I see why you want to make the API change. I will try that.

Reply:
Hmm, I might need your help to look at this closer.  When I make the change you suggest I get this error:

```
 ------ ----------------------------------------------------------------------------------------------------------
  Line   lib/Doctrine/DBAL/Portability/Connection.php
 ------ ----------------------------------------------------------------------------------------------------------
  114    Method Doctrine\DBAL\Portability\Connection::prepare() should return Doctrine\DBAL\Statement but returns
         Doctrine\DBAL\Portability\Statement.
 ------ ----------------------------------------------------------------------------------------------------------
```

It opens up a can of worms I am not educated enough on to make decisions on what to do. The portability stuff is new since I last worked on DBAL.

Reply:
It will require the segregation of the DBAL interfaces from the driver ones. I had this in mind for a long time and this is one more reason to do that.

Please suppress this failure in the PHPStan config instead of fixing it. The API is wrong.

Reply:
Done.

Reply:
I've filed #3591.